### PR TITLE
Fix 44 http retry

### LINKED ISSUE #44: Enhance Execute HTTP task with retrial capabilities
**Is your request related to a problem? Please describe.**
There are cases when the flow needs a retry capability of the Execute HTTP Task when a specific HTTP call response is received.
For example if the task receives a HTTP 504, the flow might want to retry for a number of times before giving-up on the call.

**Describe the solution you'd like**
Allow the end-user to select an HTTP list of error codes for which the the task if receives the http response code it will retry for a number of x times - again configurable in the UI.

**Describe the benefits or justification for this request**
Sometime calls might fail due to network reasons, a retry mechanism helps in getting the workflow with success to its end.

**Additional context**

### PR DESCRIPTION
**Pull request by florentinvintila**
Closes #44 

Introducing new functionality to retry requests for a specific number of times in case of fails.

Updated URL use for deprecated use.
Updated dependencies for new library versions.

#### Changelog

**New**

- added new parameters for task input :errorcodes = "", httperrorretry = 5

**Changed**

- N/A

**Removed**

- N/A

#### Testing / Reviewing

Testes into DEV environment and local server.

### CODE REVIEW

File: commands/http.js [hunk 1]
```diff
@@ -20,7 +21,7 @@ module.exports = {
 
     //Destructure and get properties ready.
     const taskProps = utils.resolveInputParameters();
-    const { url, method, header, contentType, body, allowUntrustedCerts, outputFilePath } = taskProps;
+    const { url, method, header, contentType, body, allowUntrustedCerts, outputFilePath, errorcodes = "", httperrorretry = 5 } = taskProps;
```
Comment (tlawrie) [inline commands/http.js:24]: One nitpick here. The default for error retries should probably match the existing (or non existent) functionality i.e. be 0. And then changed as needed by end user.

Also, make sure to match with a default in the task template.

File: commands/HTTPRetryRequest.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  return new Promise((resolve, reject) => {
+    let requestInstance = https.request(options, response => {
+      const innerStatusCode = response.statusCode.toString();
+      const responseInstance = response
+        .on("error", error => {
+          responseInstance.abort();
+          if (_self.config.ERROR_CODES && _self.config.ERROR_CODES.test(innerStatusCode) && _self.config.retryCount < _self.config.MAX_RETRIES) {
+            // if the status is one of the ones we want to retry, then make the same request
+            resolve(new HTTPRetryRequest(_self.config, _self.options));
+          } else {
```
Comment (tlawrie) [inline commands/HTTPRetryRequest.js:36]: From my understanding of the issue, the error codes are not related to retries. Its different functionality.

Error codes was meant to be, if you want the task to fail on certain response codes. As right now it only fails the task if something goes wrong with making the call, and then you have to run another task after it to check the response code resultParameter.

**Review by tlawrie — CHANGES_REQUESTED** [review]
Review comments provided, please adjust or discuss before merging.

File: commands/HTTPRetryRequest.js (hunk 2 above)
Comment (florentinvintila) [reply commands/HTTPRetryRequest.js:36]: @tlawrie how it was described to me it was to retry only when the error codes that are inputed are found as the status code. If your understanding is different please append the task with all the details and I will do the changes.

File: commands/http.js (hunk 1 above)
Comment (florentinvintila) [reply commands/http.js:24]: In case the errorcodes is an empty string the number of retires are ignored.
Will do the change as suggested.

File: commands/HTTPRetryRequest.js (hunk 2 above)
Comment (tlawrie) [reply commands/HTTPRetryRequest.js:36]: Hi @florentinvintila interesting ok. We may need to clarify with @morarucostel 

As far as I know we had two different tasks;

**1. [Cause the task to fail if a certain Status Code range is not achieved](https://github.com/boomerang-io/roadmap/issues/34)**
i.e. 2xx / 3xx / 4xx / 5xx then fail the task. This would most likely be a drop down of Status Code ranges. OR the flip of this. A drop down of what is considered success, i.e. the task only succeeds if a 2xx is received. This could also be a check box that states "Task only succeeds on 2xx Status Code" or something like this.

**2. [Retry the call y number of times](https://github.com/boomerang-io/worker.flow/issues/44)**
Add the ability to specify a certain number of retries, utilize if a timeout is received.

They aren't mutually related and in fact two separate outcomes. However the last time it was discussed, if both changes were done at the same time, then the testing would be less overall.

Comment (morarucostel) [reply commands/HTTPRetryRequest.js:36]: Hi @tlawrie , you are correct, there are two mutually (at least they started as) unrelated requirements but now that I look at them, we might need to treat them as one. Treating them separately will raise some edge cases that are difficult to manage and will become more difficult for the end-user to configure them. 
Requirement number 2, from your list, is the one that is fully addressed in this PR, but now that we think of, it doesn't make sense to have retry capability when the status code is considered a `success`. 

For that reason, I think we should address both requirements together by: 
 - using a single input text field with the Status codes that are mapping the "Error" scenario. We can allow the end-user to put "4xx,5xx" or even more granular error codes "4xx,502,503". 
 - and using a second input number field with the number of retries in case the HTTP status code is coming as one of the listed values mapped as error. 
This way we keep to a minimum the number of input fields the end-user has to work with, there are no edge cases of overlapping different types of codes and we support both requirements.

Comment (morarucostel) [reply commands/HTTPRetryRequest.js:36]: @tlawrie We had another call today with @florentinvintila and agreed on the following:
We will use three fields in order to model the solution:
 - using an input text field with the Status codes that are defined for the scenario. We can allow the end-user to put "4xx,5xx" or even more granular error codes "4xx,502,503".
 - using a drop down with the options Success or Failure, that defines the scenario selected by the end-user.
 - using a third field, input number with the number of retries in case the HTTP status code is coming as an error mapped one.

That means:
- if the end-user select a range of 2xx, `success` in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it doesn't receive a 2xx http code). The task will end with failure (in the scenario where no 2xx is received within those 3 retries),
- if the end-user select a range of 4xx,5xx, `failure` in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it receives a 4xx or 5xx http code). If it receives a return code in the 2xx area, it will finish and return as a successful run. Anything else (other than 4xx, 5xx and 2xx) is not retried and is closed as failed run. 
This model is covering the edge cases of custom defined api statuses by various implementations.

Comment (tlawrie) [reply commands/HTTPRetryRequest.js:36]: Ok. That makes sense. And thanks for going through the use detail.

However, it does present a problem where this is now really complex. With three combinations of things that a user has to adjust it becomes quite tedious, particularly for the basic users. 

I am thinking we could pair it back a bit...
1. Create a separate task of Advanced HTTP Task that has all this extra logic and they share some of the same code under the covers
2. Do we really need the custom range or can we not just have a drop-down with 1xx, 2xx, 3xx, 4xx, 5xx?

**Comment by tlawrie** [comment]
Also another review comment, the `commands/` folder is used by the worker to dynamically handle tasks etc that are passed in.

Unless the `HTTPRetryRequest.js` is needing to be a task in itself it probably needs to go in another folder OR just be a private method in the `http.js`

File: commands/HTTPRetryRequest.js (hunk 2 above)
Comment (florentinvintila) [reply commands/HTTPRetryRequest.js]: @tlawrie please take another look at it. An additional delay time was introduced between retries.

Comment (tlawrie) [reply commands/HTTPRetryRequest.js]: Thanks @florentinvintila so I think we are going to have to split this out as a separate task. They can both share the same base HTTP execution but this is now super complex.

For a standard user having to come and use this task and worry about what to see and these multiple combinations of status codes and status options. 

I think we need to have an Execute `Advanced HTTP Call` because we now are taking in 4 extra options that a standard not technical no-code user is not going to want to have to deal with.

Also what happened to simplifying the error codes
> Do we really need the custom range or can we not just have a drop-down with 1xx, 2xx, 3xx, 4xx, 5xx?

Comment (morarucostel) [reply commands/HTTPRetryRequest.js]: Hey @tlawrie , creating a new Advanced HTTP task will complicates a lot our uses case because we have many existing workflows relying on the current `Execute HTTP Call` and we rely heavily on the backward compatibility. If we introduce a new task, we'll have to change all our existing workflows and introduce a lot of regression. 

Instead we want to simplify the experience by:
 - setting the defaults (of these new 3 fields) in such a way that a typical "business" user doesn't even have to change or look at them (allowing backward compatibility),
 - basically the end-user doesn't need to do anything in that section and it will run just as before, so we don't introduce complexity.

Comment (marcusdroy) [reply commands/HTTPRetryRequest.js]: @tlawrie @morarucostel  perhaps our existing one can be renamed to Advanced HTTP Call, then we create a new basic HTTP task?

Comment (morarucostel) [reply commands/HTTPRetryRequest.js]: @marcusdroy That is great idea. Just renaming the current task to Advanced HTTP Call means the existing workflows will continue to use the same task based on the id of the task. 
I believe this might work.

**Review by morarucostel — APPROVED** [review]
Looks very good to me, Thank you @florentinvintila

**Comment by morarucostel** [comment]
@tlawrie , @marcusdroy  +1 ?

File: commands/HTTPRetryRequest.js (hunk 2 above)
Comment (tlawrie) [reply commands/HTTPRetryRequest.js]: Happy for a new task. I personally didn't want two tasks but at the same time, i don't think the number and logic of the options is going to be easy for a non-technical user to understand. Its quite the conundrum.

So i think if we go with the `Execute Advanced HTTP Call` and `Execute Basic HTTP Call` how do we go about ensuring that the code is reused as much as possible? does the task template for the Basic call just call the same method but doesn't pass in some of the inputs and hence they are defaulted?

**Comment by tlawrie** [comment]
Thanks @morarucostel for the prompt. I have commented in the thread around closing out the code change.

Also is there a screenshot of the new Task Templates and what / how the new input questions are phrased?

**Comment by morarucostel** [comment]
@tlawrie yes, this is a screenshot with the `Advanced Execute HTTP Call` new section. The `Basic Execute HTTP Call` remains the same.
image

**Comment by tlawrie** [comment]
@morarucostel thanks for the screenshot. 

1. Are the error codes just for retries? i thought they were in conjunction with the status and hence it would fail on 500 for example.
2. Maybe add a hint to the error codes such as "Comma-delimited list of error codes. You can use exact or (however you describe 5xx)"

Overall, happy with whats going on. Can you put together the new Task Template for `Execute Basic HTTP Call`

**Comment by tlawrie** [comment]
Based on community meet up, @gchickma to check with @morarucostel if this will make it into 3.6.1 closing by end of week along with the loader changes required to update the current Task Template, creation of new Task Template, and the defualt worker version change

**Comment by morarucostel** [comment]
Hi @tlawrie ,

>     1. Are the error codes just for retries? i thought they were in conjunction with the status and hence it would fail on 500 for example.
No, the codes are not just for retries they are also for determining the status of the task. The description of the execution:
>  - if the end-user select a range of 2xx, success in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it doesn't receive a 2xx http code). The task will end with failure (in the scenario where no 2xx is received within those 3 retries),
>  - if the end-user select a range of 4xx,5xx, failure in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it receives a 4xx or 5xx http code). If it receives a return code in the 2xx area, it will finish and return as a successful run. Anything else (other than 4xx, 5xx and 2xx) is not retried and is closed as failed run.
    This model is covering the edge cases of custom defined api statuses by various implementations.
 
>     2. Maybe add a hint to the error codes such as "Comma-delimited list of error codes. You can use exact or (however you describe 5xx)".
Well, the description of `Comma-delimited list of error codes` wouldn't be correct because the codes don't represent error codes, based on the drop-down from the following field where the end user can select Success or Failure they can also represent Success. Thus the reason of remaining within the `HTTP response codes` boundaries. 

Yes, I can put together a Task Template for Execute Basic HTTP Call.

**Comment by morarucostel** [comment]
OK with moving this version to the loader?
